examples/resadapter/train_sd_resadapter.py

```diff
@@ -0,0 +1,1115 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Training script for ResAdapter: Domain Consistent Resolution Adapter for Diffusion Models
+# Paper: https://arxiv.org/abs/2403.02084
+# Code: https://github.com/bytedance/res-adapter
+# Project Page: https://res-adapter.github.io/
+# Authors: [Jiaxiang Cheng](https://github.com/jiaxiangc), Pan Xie, Xin Xia, Jiashi Li, Jie Wu, Yuxi Ren, Huixia Li, Xuefeng Xiao, Min Zheng, Lean Fu
+# Training script by [Aryan V S](https://github.com/a-r-r-o-w)
+
+import argparse
+import logging
+import math
+import os
+import random
+import shutil
+from pathlib import Path
+
+import datasets
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torch.utils.checkpoint
+import transformers
+from accelerate import Accelerator
+from accelerate.logging import get_logger
+from accelerate.utils import ProjectConfiguration, set_seed
+from datasets import load_dataset
+from huggingface_hub import create_repo, upload_folder
+from packaging import version
+from peft import LoraConfig
+from peft.utils import get_peft_model_state_dict
+from torch.nn import GroupNorm
+from torch.utils.data import BatchSampler, DataLoader
+from torchvision import transforms
+from tqdm.auto import tqdm
+from transformers import CLIPTextModel, CLIPTokenizer
+
+import diffusers
+from diffusers import AutoencoderKL, DDPMScheduler, DiffusionPipeline, StableDiffusionPipeline, UNet2DConditionModel
+from diffusers.models.resnet import ResnetBlock2D
+from diffusers.optimization import get_scheduler
+from diffusers.training_utils import cast_training_params, compute_snr
+from diffusers.utils import check_min_version, convert_state_dict_to_diffusers, is_wandb_available
+from diffusers.utils.hub_utils import load_or_create_model_card, populate_model_card
+from diffusers.utils.import_utils import is_xformers_available
+from diffusers.utils.torch_utils import is_compiled_module
+
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.27.0.dev0")
+
+logger = get_logger(__name__, log_level="INFO")
+
+
+class SizeBatchSampler(BatchSampler):
```

Comment:
@jiaxiangc Here are the relevant changes I've made to the lora training script for supporting ResAdapter training.

1. SizeBatchSampler allows for the creation of various image size buckets, which are then shuffled randomly. This is equivalent to the probability function that you describe I believe.